packages/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java

```diff
@@ -484,3 +493,41 @@ public void onFailure(@NonNull Exception e) {
     }
   }
 }
+
+final class FirestoreMessageCodec extends StandardMessageCodec {
+  public static final FirestoreMessageCodec INSTANCE = new FirestoreMessageCodec();
+  private static final byte DATE_TIME = (byte) 128;
+  private static final byte GEO_POINT = (byte) 129;
+  private static final byte DOCUMENT_REFERENCE = (byte) 130;
+
+  @Override
+  protected void writeValue(ByteArrayOutputStream stream, Object value) {
+    if (value instanceof Date) {
+      stream.write(DATE_TIME);
+      writeLong(stream, ((Date) value).getTime());
+    } else if (value instanceof GeoPoint) {
+      stream.write(GEO_POINT);
+      writeValue(stream, ((GeoPoint) value).getLatitude());
+      writeValue(stream, ((GeoPoint) value).getLongitude());
```

Comment:
You could use `writeDouble` here twice to avoid the redundant type discriminator byte you get with each call to `writeValue`. That would of course require changing the corresponding reader as well as the Dart and iOS sides too.